File: wpilibjExamples/src/main/java/edu/wpi/first/wpilibj/examples/commandbasedbestpracticeled/AddressableLEDBuffer.java
// Copyright (c) FIRST and other WPILib contributors.
// Open Source Software; you can modify and/or share it under the terms of
// the WPILib BSD license file in the root directory of this project.

package frc.robot;

// import edu.wpi.first.wpilibj.util.Color;
// import edu.wpi.first.wpilibj.util.Color8Bit;

/** Buffer storage for Addressable LEDs. */
public class AddressableLEDBuffer implements LEDReader, LEDWriter {
  byte[] m_buffer;

  /**
   * Constructs a new LED buffer with the specified length.
   *
   * @param length The length of the buffer in pixels
   */
  public AddressableLEDBuffer(int length) {
    m_buffer = new byte[length * 4];
  }

  /**
   * Sets a specific led in the buffer.
   *
   * @param index the index to write
   * @param r the r value [0-255]
   * @param g the g value [0-255]
   * @param b the b value [0-255]
   */
  @Override
  public void setRGB(int index, int r, int g, int b) {
    m_buffer[index * 4] = (byte) b;
    m_buffer[(index * 4) + 1] = (byte) g;
    m_buffer[(index * 4) + 2] = (byte) r;
    m_buffer[(index * 4) + 3] = 0;
  }

  /**
   * Sets a specific LED in the buffer.
   *
   * @param index The index to write
   * @param color The color of the LED
   */
  @Override
  public void setLED(int index, Color color) {
    setRGB(index, (int) (color.red * 255), (int) (color.green * 255), (int) (color.blue * 255));
  }

  /**
   * Sets a specific LED in the buffer.
   *
   * @param index The index to write
   * @param color The color of the LED
   */
  @Override
  public void setLED(int index, Color8Bit color) {
    setRGB(index, color.red, color.green, color.blue);
  }

  /**
   * Gets the buffer length.
   *
   * @return the buffer length
   */
  @Override
  public int getLength() {
    return m_buffer.length / 4;
  }

  /**
   * Gets the color at the specified index.
   *
   * @param index the index to get
   * @return the LED color at the specified index
   */
  @Override
  public Color8Bit getLED8Bit(int index) {
    return new Color8Bit(getRed(index), getGreen(index), getBlue(index));
  }

  /**
   * Gets the color at the specified index.
   *
   * @param index the index to get
   * @return the LED color at the specified index
   */
  @Override
  public Color getLED(int index) {
    return new Color(getRed(index) / 255.0, getGreen(index) / 255.0, getBlue(index) / 255.0);
  }

  /**
   * Gets the red channel of the color at the specified index.
   *
   * @param index the index of the LED to read
   * @return the value of the red channel, from [0, 255]
   */
  @Override
  public int getRed(int index) {
    return m_buffer[index * 4 + 2] & 0xFF;
  }

  /**
   * Gets the green channel of the color at the specified index.
   *
   * @param index the index of the LED to read
   * @return the value of the green channel, from [0, 255]
   */
  @Override
  public int getGreen(int index) {
    return m_buffer[index * 4 + 1] & 0xFF;
  }

  /**
   * Gets the blue channel of the color at the specified index.
   *
   * @param index the index of the LED to read
   * @return the value of the blue channel, from [0, 255]
   */
  @Override
  public int getBlue(int index) {
    return m_buffer[index * 4] & 0xFF;
  }

  /**
   * Creates a view of a subsection of this data buffer, starting from (and including) {@code
   * startingIndex} and ending on (and including) {@code endingIndex}. Views cannot be written
   * directly to an {@link AddressableLED}, but are useful tools for logically separating different
   * sections of an LED strip for independent control.
   *
   * @param startingIndex the first index in this buffer that the view should encompass (inclusive)
   * @param endingIndex the last index in this buffer that the view should encompass (inclusive)
   * @return the view object
   */
  public AddressableLEDBufferView createView(int startingIndex, int endingIndex) {
    return new AddressableLEDBufferView(this, startingIndex, endingIndex);
  }
}

File: wpilibc/src/main/python/wpilib/src/rpy/AddressableLEDBuffer.cpp
// Copyright (c) FIRST and other WPILib contributors.
// Open Source Software; you can modify and/or share it under the terms of
// the WPILib BSD license file in the root directory of this project.

#include "rpy/AddressableLEDBuffer.h"

#include <span>
#include <stdexcept>

namespace wpi {

void AddressableLEDBuffer::SetRGB(size_t index, int r, int g, int b) {
  m_buffer.at(index).SetRGB(r, g, b);
}

void AddressableLEDBuffer::SetHSV(size_t index, int h, int s, int v) {
  m_buffer.at(index).SetHSV(h, s, v);
}

void AddressableLEDBuffer::SetLED(size_t index, const wpi::util::Color& color) {
  m_buffer.at(index).SetLED(color);
}

void AddressableLEDBuffer::SetLED(size_t index,
                                  const wpi::util::Color8Bit& color) {
  m_buffer.at(index).SetLED(color);
}

int AddressableLEDBuffer::GetRed(size_t index) const {
  return m_buffer.at(index).r;
}

int AddressableLEDBuffer::GetGreen(size_t index) const {
  return m_buffer.at(index).g;
}

int AddressableLEDBuffer::GetBlue(size_t index) const {
  return m_buffer.at(index).b;
}

wpi::util::Color AddressableLEDBuffer::GetLED(size_t index) const {
  const auto& led = m_buffer.at(index);
  return wpi::util::Color{led.r / 255.0, led.g / 255.0, led.b / 255.0};
}

wpi::util::Color8Bit AddressableLEDBuffer::GetLED8Bit(size_t index) const {
  const auto& led = m_buffer.at(index);
  return wpi::util::Color8Bit{led.r, led.g, led.b};
}

AddressableLED::LEDData& AddressableLEDBuffer::at(size_t index) {
  return m_buffer.at(index);
}

AddressableLED::LEDData& AddressableLEDBuffer::operator[](size_t index) {
  return m_buffer.at(index);
}

const AddressableLED::LEDData& AddressableLEDBuffer::operator[](
    size_t index) const {
  return m_buffer.at(index);
}

void AddressableLEDBuffer::View::SetRGB(size_t index, int r, int g, int b) {
  at(index).SetRGB(r, g, b);
}

void AddressableLEDBuffer::View::SetHSV(size_t index, int h, int s, int v) {
  at(index).SetHSV(h, s, v);
}

void AddressableLEDBuffer::View::SetLED(size_t index,
                                        const wpi::util::Color& color) {
  at(index).SetLED(color);
}

void AddressableLEDBuffer::View::SetLED(size_t index,
                                        const wpi::util::Color8Bit& color) {
  at(index).SetLED(color);
}

AddressableLED::LEDData& AddressableLEDBuffer::View::at(size_t index) {
  // std::span::at doesn't exist until C++26
  if (index >= m_data.size()) {
    throw std::out_of_range("Index out of range");
  }
  return m_data[index];
}

AddressableLED::LEDData& AddressableLEDBuffer::View::operator[](size_t index) {
  return at(index);
}

const AddressableLED::LEDData& AddressableLEDBuffer::View::at(
    size_t index) const {
  // std::span::at doesn't exist until C++26
  if (index >= m_data.size()) {
    throw std::out_of_range("Index out of range");
  }
  return m_data[index];
}

const AddressableLED::LEDData& AddressableLEDBuffer::View::operator[](
    size_t index) const {
  return at(index);
}

wpi::util::Color AddressableLEDBuffer::View::GetLED(size_t index) const {
  const auto& led = at(index);
  return wpi::util::Color{led.r / 255.0, led.g / 255.0, led.b / 255.0};
}

wpi::util::Color8Bit AddressableLEDBuffer::View::GetLED8Bit(
    size_t index) const {
  const auto& led = at(index);
  return wpi::util::Color8Bit{led.r, led.g, led.b};
}

AddressableLEDBuffer::View::View(std::span<AddressableLED::LEDData> data)
    : m_data(data) {}

AddressableLEDBuffer::View AddressableLEDBuffer::CreateView(
    pybind11::slice slice) {
  size_t start = 0, stop = 0, step = 0, slicelength = 0;
  slice.compute(m_buffer.size(), &start, &stop, &step, &slicelength);
  if (step != 1) {
    throw std::out_of_range("step != 1");
  }
  if (!slicelength) {
    throw std::out_of_range("zero length view");
  }
  return View(std::span(m_buffer).subspan(start, slicelength));
}

}  // namespace wpi
